test/integration/oauth_external_test.go

```diff
@@ -0,0 +1,215 @@
+		default:
+			t.Fatalf("Unexpected request: %v", r.URL.Path)
+		}
+	}))
+	cert, err := tls.X509KeyPair(authLocalhostCert, authLocalhostKey)
+	authServer.TLS = &tls.Config{
+		Certificates: []tls.Certificate{cert},
+	}
+	authServer.StartTLS()
+	defer authServer.Close()
+	authServerURL = authServer.URL
+
+	authConfigFile, err := ioutil.TempFile("", "test.cfg")
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	defer os.Remove(authConfigFile.Name())
+	authConfigObj := kclientcmdapi.Config{
+		Clusters: map[string]*kclientcmdapi.Cluster{
+			"authService": {
+				CertificateAuthority: caFile.Name(),
+				Server:               authServer.URL + "/authenticate",
+			},
+		},
+		AuthInfos: map[string]*kclientcmdapi.AuthInfo{
+			"apiServer": {
+				ClientCertificateData: authLocalhostCert,
+				ClientKeyData:         authLocalhostKey,
+			},
+		},
+		CurrentContext: "webhook",
+		Contexts: map[string]*kclientcmdapi.Context{
+			"webhook": {
+				Cluster:  "authService",
+				AuthInfo: "apiServer",
+			},
+		},
+	}
+	if err := kclientcmd.WriteToFile(authConfigObj, authConfigFile.Name()); err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	authServerMetadataFile, err := ioutil.TempFile("", "metadata.cfg")
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	defer os.Remove(authServerMetadataFile.Name())
+	authServerMetadata, err := json.MarshalIndent(oauthutil.GetOauthMetadata(authServer.URL), "", "")
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	authServerMetadataFile.Write(authServerMetadata)
+	authServerMetadataFile.Sync()
+	authServerMetadataFile.Close()
+
+	// Get master config
+	masterOptions, err := testserver.DefaultMasterOptions()
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	defer testserver.CleanupMasterEtcd(t, masterOptions)
+
+	masterOptions.OAuthConfig = nil
+	masterOptions.ExternalOAuthConfig = &configapi.ExternalOAuthConfig{
+		MetadataFile:    authServerMetadataFile.Name(),
+		MasterPublicURL: masterOptions.MasterPublicURL,
+		AssetPublicURL:  masterOptions.MasterPublicURL + "/console/",
+	}
+	masterOptions.AuthConfig.WebhookTokenAuthenticators = []configapi.WebhookTokenAuthenticator{
+		{
+			ConfigFile: authConfigFile.Name(),
+			CacheTTL:   "10s",
+		},
+	}
+
+	// Start server
+	clusterAdminKubeConfig, err := testserver.StartConfiguredMaster(masterOptions)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	clusterAdminClientConfig, err := testutil.GetClusterAdminClientConfig(clusterAdminKubeConfig)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	anonymousConfig := rest.AnonymousClientConfig(clusterAdminClientConfig)
+	anonymousConfig.TLSClientConfig.CAData = append(anonymousConfig.TLSClientConfig.CAData, authLocalhostCert...)
+	reader := bytes.NewBufferString("user\npass")
+	accessToken, err := tokencmd.RequestToken(anonymousConfig, reader, "", "")
+	if err != nil {
+		t.Errorf("Unexpected error: %v", err)
+	}
+	if len(accessToken) == 0 {
```

Comment:
This should check to make sure the token matches.

Reply:
ok